uk_geo_utils/base_updater.py

```diff
@@ -0,0 +1,377 @@
+
+    def validate_foreign_keys(self):
+        self.stdout.write("Validating foreign keys...")
+        for constraint in self.foreign_key_constraints:
+            self.stdout.write(f"Executing: {constraint['validate_statement']}")
+            self.cursor.execute(constraint["validate_statement"])
+
+    def create_temp_table(self):
+        self.stdout.write(
+            f"Creating temp table called {self.temp_table_name}..."
+        )
+        self.cursor.execute(f"DROP TABLE IF EXISTS {self.temp_table_name};")
+        self.cursor.execute(
+            f"CREATE TABLE {self.temp_table_name} AS SELECT * FROM {self.table_name} LIMIT 0;"
+        )
+
+    def index_exists(self, index_name):
+        self.cursor.execute(f"""
+            SELECT 1 
+            FROM pg_indexes 
+            WHERE indexname = '{index_name}'
+        """)
+
+        return bool(self.cursor.fetchone())
+
+    def rename_old_table(self):
+        self.stdout.write("Renaming old table...")
+        rename_table_statement = (
+            f"ALTER TABLE {self.table_name} RENAME TO {self.table_name}_old"
+        )
+        self.stdout.write(f"Executing: {rename_table_statement}")
+        self.cursor.execute(rename_table_statement)
+        self.stdout.write("Renaming primary key...")
+        primary_key_rename_statement = f"ALTER TABLE {self.table_name}_old RENAME CONSTRAINT {self.primary_key_constraint['constraint_name']} TO {self.primary_key_constraint['constraint_name']}_old"
+        self.stdout.write(f"Executing: {primary_key_rename_statement}")
+        self.cursor.execute(primary_key_rename_statement)
+
+        index_rename_statements = []
+
+        for index in self.indexes:
+            if self.index_exists(index["index_name"]):
+                index_rename_statements.append(
+                    index["rename_old_index_statement"]
+                )
+
+        for statement in index_rename_statements:
+            self.stdout.write(f"Executing: {statement}")
+            self.cursor.execute(statement)
+
+    def rename_temp_table(self):
+        self.stdout.write("Renaming temp table...")
+        rename_table_statement = (
+            f"ALTER TABLE {self.temp_table_name} RENAME TO {self.table_name}"
+        )
+        self.stdout.write(f"Executing: {rename_table_statement}")
+        self.cursor.execute(rename_table_statement)
+        self.stdout.write("Renaming primary key...")
+        primary_key_rename_statement = f"ALTER TABLE {self.table_name} RENAME CONSTRAINT {self.primary_key_constraint['temp_name']} TO {self.primary_key_constraint['constraint_name']}"
+        self.stdout.write(f"Executing: {primary_key_rename_statement}")
+        self.cursor.execute(primary_key_rename_statement)
+
+        index_rename_statements = []
+        for index in self.indexes:
+            if self.index_exists(index["temp_index_name"]):
+                index_rename_statements.append(
+                    index["rename_temp_index_statement"]
+                )
```

Comment:
Could maybe use `IF EXISTS` in the statement.

Reply:
I actually reckon if we can avoid manipulating the SQL statements as strings that is probably a good thing.

Follow up question: Is there a case where `temp_index_name` doesn't exist, and that's ok, or are we always expecting this to exist.

i.e: should this really be:

```py
if self.index_exists(index["temp_index_name"]):
    # do the things
else:
    raise Exception("unexpected thing happened - roll back the transaction!")
```

??

Reply:
Notes from call: If we can make this conditional on whether or not its the PK, that's ideal. If not, lets leave a comment explaining the `index_exists()` here

Reply:
Now comparing to primary key name: b4e5765cc6aac6